crates/ty_python_semantic/src/types/constraints.rs

```diff
@@ -484,6 +497,39 @@ impl<'db> ConstrainedTypeVar<'db> {
         debug_assert_eq!(lower, lower.bottom_materialization(db));
         debug_assert_eq!(upper, upper.top_materialization(db));
 
+        // It's not useful for an upper bound to be an intersection type, or for a lower bound to
+        // be a union type. Both of those can be rewritten as simpler BDDs:
+        //
+        //   T ≤ α & β  ⇒ (T ≤ α) ∧ (T ≤ β)
+        //   T ≤ α & ¬β ⇒ (T ≤ α) ∧ ¬(T ≤ β)
+        //   α | β ≤ T  ⇒ (α ≤ T) ∧ (β ≤ T)
```

Comment:
You write these as implications, but don't we need those to be equivalences?

The one about negations looks interesting. I guess it could be simplified to `T ≤ ¬β ⇒ ¬(T ≤ β)`, since the intersection part of it is already covered by the first rule. But more importantly, is this really correct? The left hand side seems to always be true for `T = Never`, whereas the right hand side seems to always be false for `T = Never`. That would mean it's not even true as an implication?